pkg/gpu/cuda/fatbin.go

```diff
@@ -0,0 +1,475 @@
+		}
+	}
+
+	return fatbin, nil
+}
+
+// nvInfoFormat defines the format (size) of a value of a .nv.info item
+type nvInfoFormat uint8
+
+const (
+	nviFmtNone nvInfoFormat = 0x01
+	nviFmtBval nvInfoFormat = 0x02
+	nviFmtHval nvInfoFormat = 0x03
+	nviFmtSval nvInfoFormat = 0x04
+)
+
+// nvInfoAttr defines the attribute of a .nv.info item
+type nvInfoAttr uint8
+
+const (
+	nviAttrError nvInfoAttr = iota
+	nviAttrPad
+	nviAttrImageSlot
+	nviAttrJumptableRelocs
+	nviAttrCtaidzUsed
+	nviAttrMaxThreads
+	nviAttrImageOffset
+	nviAttrImageSize
+	nviAttrTextureNormalized
+	nviAttrSamplerInit
+	nviAttrParamCbank
+	nviAttrSmemParamOffsets
+	nviAttrCbankParamOffsets
+	nviAttrSyncStack
+	nviAttrTexidSampidMap
+	nviAttrExterns
+	nviAttrReqntid
+	nviAttrFrameSize
+	nviAttrMinStackSize
+	nviAttrSamplerForceUnnormalized
+	nviAttrBindlessImageOffsets
+	nviAttrBindlessTextureBank
+	nviAttrBindlessSurfaceBank
+	nviAttrKparamInfo
+	nviAttrSmemParamSize
+	nviAttrCbankParamSize
+	nviAttrQueryNumattrib
+	nviAttrMaxregCount
+	nviAttrExitInstrOffsets
+	nviAttrS2rctaidInstrOffsets
+	nviAttrCrsStackSize
+	nviAttrNeedCnpWrapper
+	nviAttrNeedCnpPatch
+	nviAttrExplicitCaching
+	nviAttrIstypepUsed
+	nviAttrMaxStackSize
+	nviAttrSuqUsed
+	nviAttrLdCachemodInstrOffsets
+	nviAttrLoadCacheRequest
+	nviAttrAtomSysInstrOffsets
+	nviAttrCoopGroupInstrOffsets
+	nviAttrCoopGroupMaxRegids
+	nviAttrSw1850030War
+	nviAttrWmmaUsed
+	nviAttrHasPreV10Object
+	nviAttrAtomf16EmulInstrOffsets
+	nviAttrAtom16EmulInstrRegMap
+	nviAttrRegcount
+	nviAttrSw2393858War
+	nviAttrIntWarpWideInstrOffsets
+	nviAttrSharedScratch
+	nviAttrStatistics
+
+	// New between cuda 10.2 and 11.6
+	nviAttrIndirectBranchTargets
+	nviAttrSw2861232War
+	nviAttrSwWar
+	nviAttrCudaApiVersion
+	nviAttrNumMbarriers
+	nviAttrMbarrierInstrOffsets
+	nviAttrCoroutineResumeIdOffsets
+	nviAttrSamRegionStackSize
+	nviAttrPerRegTargetPerfStats
+
+	// New between cuda 11.6 and 11.8
+	nviAttrCtaPerCluster
+	nviAttrExplicitCluster
+	nviAttrMaxClusterRank
+	nviAttrInstrRegMap
+)
+
+// nvInfoItem is the in-file representation of an item  header in the .nv.info section. The value follows
+// according to the format
+type nvInfoItem struct {
+	Format nvInfoFormat
+	Attr   nvInfoAttr
+}
+
+// cubinParser is a helper struct to parse the cubin ELF sections
+type cubinParser struct {
+	kernels            map[string]*CubinKernel
+	sectPrefixToParser map[string]func(*elf.Section, *bytes.Reader, string) error
+}
+
+func newCubinParser() cubinParser {
+	cp := cubinParser{
+		kernels:            make(map[string]*CubinKernel),
+		sectPrefixToParser: make(map[string]func(*elf.Section, *bytes.Reader, string) error),
+	}
+
+	cp.sectPrefixToParser[".nv.info"] = cp.parseNvInfoSection
+	cp.sectPrefixToParser[".text"] = cp.parseTextSection
+	cp.sectPrefixToParser[".nv.shared"] = cp.parseSharedMemSection
+	cp.sectPrefixToParser[".nv.constant"] = cp.parseConstantMemSection
+
+	return cp
+}
+
+func (cp *cubinParser) getKernel(name string) *CubinKernel {
```

Comment:
what is the difference between this function and `GetKernel` of the fatbin struct?

Reply:
This one gets the kernel for a single cubin payload, the one in `Fatbin` gets the kernel from all the cubin's that are embedded in the binary.

Reply:
i think this function should be renamed to `getOrCreateKernel`